federation/normalize.go

```diff
@@ -150,8 +150,14 @@ func mergeSameAlias(selections []*graphql.Selection) ([]*graphql.Selection, erro
 		if last == nil || selection.Alias != last.Alias {
 			// Make a copy of the selection so we can modify it below
 			// or when we flatten recursively later.
-			copy := *selection
-			selection = &copy
+			cp := *selection
+			if selection.SelectionSet != nil {
+				// Make a new SelectionSet for the copy so we will not append to the original slice later.
+				cp.SelectionSet = &graphql.SelectionSet{}
+				cp.SelectionSet.Selections = append(cp.SelectionSet.Selections, selection.SelectionSet.Selections...)
+				cp.SelectionSet.Fragments = append(cp.SelectionSet.Fragments, selection.SelectionSet.Fragments...)
```

Comment:
THanks for including the benchmark!
From it, I see that this increases runtime noticeably. In thinking about how we might lower it, I wonder how often the code creates a copy but doesn't append to one or both of the selectionsets? If that happens frequently, would it be worthwhile to only copy when the code first wants to append? We would need to track whether a copy needs to be made to permit modification, but that might avoids the allocations.

Reply:
Yeah, I tried to do this and below is the benchmark: 

```
BenchmarkNormalize2-16    	    4471	    235298 ns/op	  102405 B/op	    2132 allocs/op
```

I have to add a map[*graphql.SelectionSet]struct{} to as the input of `flatten` and `mergeSameAlias` to track whether we need to shallow copy a certain selectionSet across the recursions. 

The allocation is similar as the original and the allocation per is slightly higher.

Reply:
I updated the commit!